doc/spacebrew-midi.1
.TH "SPACEBREW\-MIDI" "" "February 2014" "" ""
.SH "NAME"
\fBspacebrew-midi\fR
.P
Route MIDI events via Spacebrew
.SH CLI Utility Usage
.P
The CLI utility is a quick and easy way to connect MIDI devices to the public Spacebrew admin interface\.
.P
Input ports start listening for two Note On events in the background to use as lower and upper boundaries for the range\. The UI doesn\'t currently indicate this is occuring—sorry, bad UX! I\'ll get that fixed in the next version\.
.P
Output ports are currently connected for the entire C\-0 through G\-9 range\.
.P
WARNING: Simultaneous connections of the same type (input or output) segfaults on some platforms\. Looking into the root cause\.\.\.
.P
.RS 2
.EX
$ npm i \-g spacebrew\-midi
$ spacebrew\-midi

Spacebrew MIDI \- Select a device to connect to Spacebrew                    
  \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-  
   INPUT on QUNEO:0                                                         
   INPUT on MPK mini:0                                                      
  OUTPUT on QUNEO:0 \- Connected!                                            
  OUTPUT on MPK mini:0
.EE
.RE
.SH Module Usage Example
.P
.RS 2
.EX
var sbmidi = require("spacebrew\-midi");

// Connect to Spacebrew, passing an onOpen handler
sbmidi\.connect(function () {

  // Determine the last available port, usually the most recently connected device
  var portNumber = sbmidi\.getInPortCount() \- 1;

  // One octave to either side of Middle C
  sbmidi\.addInputRange(portNumber, 48, 72);

  // Similar for output!
  sbmidi\.addOuputRange(portNumber, 48, 72);

  // Alternatively, just pass a port number and bounds will be "learned"
  // based on the next two Note On events received
  sbmidi\.addInputRange(portNumber);

});
.EE
.RE
.P
And over in the Admin interface you\'ll see:
.P
Screenshot of Spacebrew Admin running example code \fIhttps://github\.com/randallagordon/spacebrew\-midi/raw/master/img/readme\-example\.png\fR
.P
By default this will create a connection to the public Spacebrew
Admin \fIhttp://spacebrew\.github\.com/spacebrew/admin/admin\.html?server=sandbox\.spacebrew\.cc\fR
interface\.
.P
To connect to another server: \fBsbmidi\.connect({ server: "http://localhost/" })\fR
.P
To use a different name and description: \fBsbmidi\.connect({ name: "Bob", description: "Bob\'s MIDI Controller" })\fR
.SH CHANGELOG
.SS v0\.2\.0
.RS 0
.IP \(bu 2
Added CLI utility
.IP \(bu 2
Added support for connecting multiple ports simultaneously (still has issues)
.IP \(bu 2
Added input range addition via "learning" from MIDI Note On events

.RE
.SS v0\.1\.1
.RS 0
.IP \(bu 2
Added MIDI output support

.RE
.SH TODO
.RS 0
.IP \(bu 2
Support more MIDI messages than just note on/off
.IP \(bu 2
Use Spacebrew custom messages for sending complex CC messages (knobs, sliders, etc\.)
.IP \(bu 2
Publish MIDI helper library separately
.IP \(bu 2
Add CLI option parsing for non\-interactive routing
.IP \(bu 2
JSON based config files to support building a device map library (think \fBspacebrew\-midi \-\-device QuNeo\fR)
.IP \(bu 2
Web interface to complement CLI for adding/editing routed messages
.IP \(bu 2
Handle closing things down (onClose handler to close MIDI ports, etc\.)
.IP \(bu 2
Refactor so a range can be added on an already open port

.RE
.SH LICENSE
.P
MIT



lib/spacebrew-midi.js
/*
 * spacebrew-midi
 * https://github.com/randallagordon/spacebrew-midi
 *
 * Copyright (c) 2013-2014 Randall A. Gordon <randall@randallagordon.com>
 * Licensed under the MIT license.
 */

"use strict";

var Spacebrew = require("spacebrew.js").Spacebrew,
    midi = require("midi"),
    midiUtil = require("./midiutil.js");

var sb,
    midiIns = {},
    midiOuts = {};

var defaults = {
  server: "sandbox.spacebrew.cc",
  name: "MIDI -> Spacebrew",
  description: "MIDI events, on Spacebrew! Code on GitHub: https://github.com/randallagordon/spacebrew-midi"
};

// MIDI Helpers
function getPorts(midiObj) {
  var i, ports = [];

  for (i = 0; i < midiObj.getPortCount(); i++) {
    ports.push(midiObj.getPortName(i));
  }

  return ports;
}

module.exports.getInPorts = function() {
  return getPorts(new midi.input());
};

module.exports.getOutPorts = function() {
  return getPorts(new midi.output());
};

module.exports.getInPortCount = function() {
  return (new midi.input()).getPortCount();
};

module.exports.getOutPortCount = function() {
  return (new midi.output()).getPortCount();
};


// Handlers
var handleSB = function(name, value) {
  var message = name.split(" - "),
      portName = message[0],
      note = message[1],
      hex = midiUtil.dehumanizeNote(note);

  console.log("Spacebrew Message - Name: %s\tValue: %s\tHex: %s", name, value, (+hex).toString(16));

  if (value) {
    midiOuts[portName].sendMessage([0x90, hex, 0x7F]);
  } else {
    midiOuts[portName].sendMessage([0x80, hex, 0x7F]);
  }
};

var handleMIDI = function(portName, deltaTime, message) {
  switch (true) {
  case (message[0] >= 0x80 && message[0] <= 0x8F): // Note OFF
    console.log("%s - Note off: %s\t\t\t\t[ %s, %s, %s ]", portName, midiUtil.humanizeNote(message[1]), message[0], message[1], message[2]);

    if (midiIns[portName].mode === "latch") {
      // Do nothing
    } else {
      sendSB(false);
    }

    break;
  case (message[0] >= 0x90 && message[0] <= 0x9F && message[2] > 0): // Note ON, Velocity > 0
    console.log("%s - Note on:  %s with velocity 0x%s\t[ %s, %s, %s ]", portName, midiUtil.humanizeNote(message[1]), message[2].toString(16), message[0], message[1], message[2]);


    if (midiIns[portName].mode === "latch") {
      midiIns[portName].state = !midiIns[portName].state;
      sendSB(midiIns[portName].state);
    } else {
      sendSB(true);
    }

    break;
  default:
    //console.log( "m:",  message, " d:", deltaTime );
  }

  function sendSB(value) {
    console.log("Sending ", value);
    sb.send(portName + " - " + midiUtil.humanizeNote(message[1]), "boolean", value.toString());
  }
};


// Spacebrew Functions
module.exports.connect = function(done, opts) {
  opts = opts || {};

  sb = new Spacebrew.Client(opts.server || defaults.server,
    opts.name || defaults.name,
    opts.description || defaults.description);

  sb.onBooleanMessage = handleSB;
  sb.onOpen = done;
  sb.connect();
};

module.exports.addInputRange = function(port, lowerBound, upperBound, mode) {
  var midiIn = new midi.input();
  var portName = midiIn.getPortName(port);
  mode = mode || "passthrough";

  midiIn.openPort(port);
  midiIns[portName] = { midi: midiIn, mode: mode, state: false };

  // If bounds are undefined, "learn" them
  if (typeof lowerBound === "undefined") {
    learnBounds();
  } else {
    finishSetup();
  }

  function learnBounds() {
    console.log("Listening for lower bound note on %s...", portName);
    midiIn.on("message", function learnBoundsHandler(dT, message) {
      // Note On messages with velocity > 0
      if (message[0] >= 0x90 && message[0] <= 0x9F && message[2] > 0) {
        if (typeof lowerBound === "undefined") {
          lowerBound = message[1];
          console.log("Listening for upper bound note on %s...", portName);
        } else  {
          upperBound = message[1];
          midiIn.removeListener("message", learnBoundsHandler);
          finishSetup();
        }
      }
    });
  }

  function finishSetup() {
    console.log("Connecting %s from %s to %s", portName, midiUtil.humanizeNote(lowerBound), midiUtil.humanizeNote(upperBound));

    midiIn.on("message", function (deltaTime, message) {
      return handleMIDI(portName, deltaTime, message);
    });

    for (var i = lowerBound; i <= upperBound; i++) {
      sb.addPublish(portName + " - " + midiUtil.humanizeNote(i), "boolean", "false");
    }
  }
};

module.exports.addOutputRange = function(port, lowerBound, upperBound) {
  var note;

  var midiOut = new midi.output();
  var portName = midiOut.getPortName(port);
  midiOut.openPort(port);
  midiOuts[portName] = midiOut;

  for (var i = lowerBound; i <= upperBound; i++) {
    note = midiUtil.humanizeNote(i);

    if (note === "...") {
      sb.addSubscribe(portName + " - " + i.toString(16), "boolean");
    } else {
      sb.addSubscribe(portName + " - " + note, "boolean");
    }
  }
};
